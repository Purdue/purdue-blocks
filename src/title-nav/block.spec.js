// title-nav test
jest.setTimeout(30000)
import {
    createNewPost,
    enablePageDialogAccept,
    getEditedPostContent,
    insertBlock,
} from '@wordpress/e2e-test-utils';

import {
    clickElementByText,
    blockStartup,
    updateRangeInput,
    clickCheckbox,
    openSidebarPanelWithTitle
} from '../test-helpers'

const block = {blockTitle: 'Titled Navigation', blockName: 'purdue-blocks/title-nav'}

describe( 'ðŸ”¬ Titled Navigation Block', () => {
    beforeAll( async () => {
        await enablePageDialogAccept();
    } );
    beforeEach( async () => {
        await createNewPost();
    } );

    test( 'Block should be available.', async () => {
        await insertBlock( 'Titled Navigation' )


        // tests that the block is properly inserted and matches the existing snapshot
        expect(await page.$('[data-type="purdue-blocks/title-nav"]')).not.toBeNull()
        expect( await getEditedPostContent() ).toMatchSnapshot()
    } )

    describe( 'ðŸ”¬ Block Editor Fields', () => {
        test( 'Richtext editor should update output correctly.', async () => {
            await blockStartup(block)

            const typeString = 'Title Test'

            // focus the title richtext editor and type in it.
            await page.focus('.block-editor-rich-text__editable.editor-title-nav__title')
            await page.keyboard.type(typeString, {delay: 10})
            
            const editedContent = await getEditedPostContent()
            
            expect( editedContent.includes(`<h2 class="pu-title-nav__title">${typeString}</h2>`)).toBe(true)
            expect( await getEditedPostContent() ).toMatchSnapshot()
        })
    })

    describe( 'ðŸ”¬ Side Panel Settings', () => {

        // Test settings in side panel
        test( 'Outline should be added correctly.', async () => {
            await blockStartup(block)
    
            //click to check the checkbox, outline should be on
            await clickCheckbox('Give Menu Items an Outline?')
            
            let editedContent = await getEditedPostContent()
            
            expect( editedContent.includes('{"hasOutline":true}')).toBe(true)
            expect( await getEditedPostContent() ).toMatchSnapshot()
            
    
            // click to uncheck the checkbox, outline should now turn off
            await clickCheckbox('Give Menu Items an Outline?')
            
            editedContent = await getEditedPostContent()
            
            expect( editedContent.includes('{"hasOutline":true}')).toBe(false)
            expect( await getEditedPostContent() ).toMatchSnapshot()
        })

        describe( 'ðŸ”¬ Adding/Removing Navigation Links', () => {

            // test settings to do with adding and removing the inner navigation links
            test( 'Navigation links added correctly.', async () => {
                await blockStartup(block)
        
                const addLinks = 3
                await updateRangeInput(`.components-input-control__input[aria-label="Number of Navigation Items"]`, addLinks)
        
                const linksNum = await page.$$eval(`[data-type="purdue-blocks/title-nav-link"]`, array => array.length)
        
                expect( linksNum ).toEqual(addLinks)
                expect( await getEditedPostContent() ).toMatchSnapshot()
        
            })
        
            test( 'Number of navigation links update correctly.', async () => {
                await blockStartup(block)
        
                let update = 5
                await updateRangeInput(`.components-input-control__input[aria-label="Number of Navigation Items"]`, update)
                
                let linksNum = await page.$$eval(`[data-type="purdue-blocks/title-nav-link"]`, array => array.length)
                expect( linksNum ).toEqual(update)
        
                update = 2
                await updateRangeInput(`.components-input-control__input[aria-label="Number of Navigation Items"]`, update)
        
        
                linksNum = await page.$$eval(`[data-type="purdue-blocks/title-nav-link"]`, array => array.length)
                expect( linksNum ).toEqual(update)
        
                expect( await getEditedPostContent() ).toMatchSnapshot()
        
            })
        })

        describe( 'ðŸ”¬ CTA Button Settings', () => {
            test( 'CTA button checkbox should toggle controls.', async () => {
                await blockStartup(block)
        
                // click the checkbox to show button controls
                await clickCheckbox('Add a CTA Button?')
                // incase the controls are not expanded by default
                await openSidebarPanelWithTitle('Button Controls')
        
                const buttonControls = await page.$('.components-panel__body.is-opened')
        
                expect(buttonControls).not.toBeNull()
        
            })
        
            test( 'CTA button controls should update output correctly.', async () => {
                await blockStartup(block)
        
                // click the checkbox to show button controls
                await clickCheckbox('Add a CTA Button?')
                // incase the controls are not expanded by default
                await openSidebarPanelWithTitle('Button Controls')
        
                const typeString = 'Button Text'
                const typeUrl = 'https://www.purdue.edu'
        
                await clickElementByText('label', 'Button Text')
                await page.keyboard.type(typeString, {delay: 10})
        
                await clickElementByText('label', 'Button Link')
                await page.keyboard.type(typeUrl, {delay: 10})
        
                const editedContent = await getEditedPostContent()
                
                expect( editedContent.includes('{"addButton":true,')).toBe(true)
                expect( editedContent.includes(`"buttonText":"${typeString}"`)).toBe(true)
                expect( editedContent.includes(`"buttonLink":"${typeUrl}"}`)).toBe(true)
                expect( await getEditedPostContent() ).toMatchSnapshot()
            })
        })
    
    })
} );
