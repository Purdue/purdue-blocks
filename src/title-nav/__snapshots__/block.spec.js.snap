// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`🔬 Titled Navigation Block 🔎 Block should be available. 1`] = `
"<!-- wp:purdue-blocks/title-nav -->
<section class=\\"section pu-title-nav \\"><h2 class=\\"pu-title-nav__title\\"></h2><ul class=\\"pu-title-nav__menu\\"><!-- wp:purdue-blocks/title-nav-link -->
<li class=\\"\\"></li>
<!-- /wp:purdue-blocks/title-nav-link --></ul></section>
<!-- /wp:purdue-blocks/title-nav -->"
`;

exports[`🔬 Titled Navigation Block 🔬 Block Editor Fields 🔎 Richtext editor should update output correctly. 1`] = `
"<!-- wp:purdue-blocks/title-nav -->
<section class=\\"section pu-title-nav \\"><h2 class=\\"pu-title-nav__title\\">Title Test</h2><ul class=\\"pu-title-nav__menu\\"><!-- wp:purdue-blocks/title-nav-link -->
<li class=\\"\\"></li>
<!-- /wp:purdue-blocks/title-nav-link --></ul></section>
<!-- /wp:purdue-blocks/title-nav -->"
`;

exports[`🔬 Titled Navigation Block 🔬 Side Panel Settings 🔎 Outline should be added correctly. 1`] = `
"<!-- wp:purdue-blocks/title-nav {\\"hasOutline\\":true} -->
<section class=\\"section pu-title-nav has-outline\\"><h2 class=\\"pu-title-nav__title\\"></h2><ul class=\\"pu-title-nav__menu\\"><!-- wp:purdue-blocks/title-nav-link -->
<li class=\\"\\"></li>
<!-- /wp:purdue-blocks/title-nav-link --></ul></section>
<!-- /wp:purdue-blocks/title-nav -->"
`;

exports[`🔬 Titled Navigation Block 🔬 Side Panel Settings 🔎 Outline should be added correctly. 2`] = `
"<!-- wp:purdue-blocks/title-nav -->
<section class=\\"section pu-title-nav \\"><h2 class=\\"pu-title-nav__title\\"></h2><ul class=\\"pu-title-nav__menu\\"><!-- wp:purdue-blocks/title-nav-link -->
<li class=\\"\\"></li>
<!-- /wp:purdue-blocks/title-nav-link --></ul></section>
<!-- /wp:purdue-blocks/title-nav -->"
`;

exports[`🔬 Titled Navigation Block 🔬 Side Panel Settings 🔬 Adding/Removing Navigation Links 🔎 Navigation links added correctly. 1`] = `
"<!-- wp:purdue-blocks/title-nav {\\"numLinks\\":3} -->
<section class=\\"section pu-title-nav \\"><h2 class=\\"pu-title-nav__title\\"></h2><ul class=\\"pu-title-nav__menu\\"><!-- wp:purdue-blocks/title-nav-link -->
<li class=\\"\\"></li>
<!-- /wp:purdue-blocks/title-nav-link -->

<!-- wp:purdue-blocks/title-nav-link -->
<li class=\\"\\"></li>
<!-- /wp:purdue-blocks/title-nav-link -->

<!-- wp:purdue-blocks/title-nav-link -->
<li class=\\"\\"></li>
<!-- /wp:purdue-blocks/title-nav-link --></ul></section>
<!-- /wp:purdue-blocks/title-nav -->"
`;

exports[`🔬 Titled Navigation Block 🔬 Side Panel Settings 🔬 Adding/Removing Navigation Links 🔎 Number of navigation links update correctly. 1`] = `
"<!-- wp:purdue-blocks/title-nav {\\"numLinks\\":2} -->
<section class=\\"section pu-title-nav \\"><h2 class=\\"pu-title-nav__title\\"></h2><ul class=\\"pu-title-nav__menu\\"><!-- wp:purdue-blocks/title-nav-link -->
<li class=\\"\\"></li>
<!-- /wp:purdue-blocks/title-nav-link -->

<!-- wp:purdue-blocks/title-nav-link -->
<li class=\\"\\"></li>
<!-- /wp:purdue-blocks/title-nav-link --></ul></section>
<!-- /wp:purdue-blocks/title-nav -->"
`;

exports[`🔬 Titled Navigation Block 🔬 Side Panel Settings 🔬 CTA Button Settings 🔎 CTA button controls should update output correctly. 1`] = `
"<!-- wp:purdue-blocks/title-nav {\\"addButton\\":true,\\"buttonText\\":\\"Button Text\\",\\"buttonLink\\":\\"https://www.purdue.edu\\"} -->
<section class=\\"section pu-title-nav \\"><div class=\\"columns\\"><div class=\\"column is-8 pu-title-nav__left\\"><h2 class=\\"pu-title-nav__title\\"></h2><ul class=\\"pu-title-nav__menu\\"><!-- wp:purdue-blocks/title-nav-link -->
<li class=\\"\\"></li>
<!-- /wp:purdue-blocks/title-nav-link --></ul></div><div class=\\"column is-3 is-offset-1 pu-title-nav__right\\"><a class=\\"pu-title-nav__cta\\" href=\\"https://www.purdue.edu\\"> Button Text </a></div></div></section>
<!-- /wp:purdue-blocks/title-nav -->"
`;
